text/3521-forbidden-function-casts.md

```diff
@@ -0,0 +1,116 @@
+- Feature Name: `forbidden_function_casts`
+- Start Date: (fill me in with today's date, 2023-11-12)
+- RFC PR: [rust-lang/rfcs#3521](https://github.com/rust-lang/rfcs/pull/3521)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to forbid all function-to-integer `as` casts in the next edition.
+
+
+# Motivation
+[motivation]: #motivation
+
+Currently we allow casting of function pointers and function item types to any integer width, returning the function address. Casts to types that are smaller than `usize` truncate, while casts to bigger types zero-extend. Thus, a function-to-int cast is basically transitive — `f as u8` is the same as `f as usize as u8`.
+
+This behavior can be surprising and error-prone. Moreover casting function item types is especially confusing, for example:
+```rust
+// This is an identity cast
+// (`u8::MAX` is a `u8` constant)
+let m = u8::MAX as u8;
+
+// This returns the least significant byte
+// of the function pointer address of `u8::max`
+// (`u8::max` is a function returning maximum of two `u8`s)
+let m = u8::max as u8
+```
+
+An example of this being a problem in practice can be seen in [rust-lang/rust#115511](https://github.com/rust-lang/rust/issues/115511):
+```rust
+// Code in `std::sys::windows::process::Command::spawn`
+si.cb = mem::size_of::<c::STARTUPINFOW> as c::DWORD;
+```
+
+Instead of casting the *result* of `size_of` to `c::DWORD` the `size_of` function itself was cast, causing issues.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In edition 2024, function-to-integer casts are a hard error:
+```rust
+let ptr: fn() = todo!();
+let addr = ptr as usize;   //~ ERROR
+let addr_byte = ptr as u8; //~ ERROR
+```
+
+To get the old behavior you can cast fn pointer to a raw pointer first and then to an integer:
+```rust
+let addr = ptr as *const () as usize;
+let addr_byte = ptr as *const () as usize as u8;
+```
+
+Those changes can be automatically applied by an edition migration lint.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+- On `edition >= 2024`
+  - Casting function pointers or function defenitions to an integer type is a hard error
+      - Note that you can still cast function pointers and function item types to `*const V` and `*mut V` where `V: Sized`
+      - Note that function item types can also be casted to function pointers
+- On `edition < 2024`
+  - Everything works as it used to
+  - In cases where we produce an error on `edition >= 2024` a FCW (Futurte Compatability Warning) is emited
```

Comment:
```suggestion
  - In cases where we produce an error on `edition >= 2024` an edition compatibility warning is emitted
```

First of all, typos. Second of all, FCW implies that a future Rust version may break the code even on old editions, which is not the desired intent here.

Reply:
Right, I was a bit confused when I was writing that 😅